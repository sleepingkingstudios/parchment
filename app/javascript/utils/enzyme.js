import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
