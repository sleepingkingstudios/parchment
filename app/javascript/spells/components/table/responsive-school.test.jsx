import React from 'react';
import { shallow } from 'enzyme';

import ResponsiveSchool from './responsive-school';

const responsiveSizes = ['xs', 'sm', 'md', 'lg', 'xl'];

const parseResponsiveStyles = (classNames) => {
  const responsiveStyles = { xs: true };

  classNames.split(' ').forEach((className) => {
    if (!className.match(/^d-/)) { return; }

    const defaultMatch = className.match(/^d-(\w+)$/);

    if (defaultMatch) {
      const [, style] = defaultMatch;

      responsiveStyles.xs = style;
    }

    const sizeMatch = className.match(/^d-(\w\w)-(\w+)/);

    if (sizeMatch) {
      const [, size, style] = sizeMatch;

      responsiveStyles[size] = style;
    }
  });

  return responsiveStyles;
};

const isVisible = (className, size) => {
  if (typeof className === 'undefined') { return true; }

  const responsiveStyles = parseResponsiveStyles(className);
  const sizeIndex = responsiveSizes.indexOf(size);

  for (let i = sizeIndex; i >= 0; i -= 1) {
    const currentSize = responsiveSizes[i];
    const style = responsiveStyles[currentSize];

    if (style) { return style !== 'none'; }
  }

  return true;
};

const responsiveText = (wrapper, size) => {
  const children = wrapper.children();

  if (children.length === 0) { return wrapper.text(); }

  if (!isVisible(wrapper.prop('className'), size)) { return ''; }

  let buffer = '';

  wrapper.children().forEach((child) => {
    buffer += responsiveText(child, size);
  });

  return buffer;
};

describe('<ResponsiveSchool />', () => {
  const school = 'pyromancy';

  describe('with level: 0', () => {
    const level = 0;

    it('should match the snapshot', () => {
      const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

      expect(rendered).toMatchSnapshot();
    });

    describe('when rendered at size: xs', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'xs')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: sm', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'sm')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: md', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'md')).toEqual(`Pyromancy ${level}`);
      });
    });

    describe('when rendered at size: lg', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'lg')).toEqual('Pyromancy cantrip');
      });
    });
  });

  describe('with level: 1', () => {
    const level = 1;

    it('should match the snapshot', () => {
      const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

      expect(rendered).toMatchSnapshot();
    });

    describe('when rendered at size: xs', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'xs')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: sm', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'sm')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: md', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'md')).toEqual(`Pyromancy ${level}`);
      });
    });

    describe('when rendered at size: lg', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'lg')).toEqual('1st-level pyromancy');
      });
    });
  });

  describe('with level: 2', () => {
    const level = 2;

    it('should match the snapshot', () => {
      const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

      expect(rendered).toMatchSnapshot();
    });

    describe('when rendered at size: xs', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'xs')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: sm', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'sm')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: md', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'md')).toEqual(`Pyromancy ${level}`);
      });
    });

    describe('when rendered at size: lg', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'lg')).toEqual('2nd-level pyromancy');
      });
    });
  });

  describe('with level: 3', () => {
    const level = 3;

    it('should match the snapshot', () => {
      const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

      expect(rendered).toMatchSnapshot();
    });

    describe('when rendered at size: xs', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'xs')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: sm', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'sm')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: md', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'md')).toEqual(`Pyromancy ${level}`);
      });
    });

    describe('when rendered at size: lg', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'lg')).toEqual('3rd-level pyromancy');
      });
    });
  });

  describe('with level: 4+', () => {
    const level = 4;

    it('should match the snapshot', () => {
      const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

      expect(rendered).toMatchSnapshot();
    });

    describe('when rendered at size: xs', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'xs')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: sm', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'sm')).toEqual(`Pyr ${level}`);
      });
    });

    describe('when rendered at size: md', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'md')).toEqual(`Pyromancy ${level}`);
      });
    });

    describe('when rendered at size: lg', () => {
      it('should display the level and school', () => {
        const rendered = shallow(<ResponsiveSchool level={level} school={school} />);

        expect(rendered).toHaveDisplayName('span');
        expect(responsiveText(rendered, 'lg')).toEqual('4th-level pyromancy');
      });
    });
  });
});
