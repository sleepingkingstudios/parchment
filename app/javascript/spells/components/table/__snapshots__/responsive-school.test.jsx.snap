// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResponsiveSchool /> with level: 0 should match the snapshot 1`] = `
<span>
  Pyr
  <span
    className="d-none d-md-inline"
  >
    omancy
  </span>
  <span
    className="d-none d-lg-inline"
  >
     cantrip
  </span>
  <span
    className="d-inline d-lg-none"
  >
     0
  </span>
</span>
`;

exports[`<ResponsiveSchool /> with level: 1 should match the snapshot 1`] = `
<span>
  <span
    className="d-none d-lg-inline"
  >
    1st
    -level 
    pyromancy
  </span>
  <span
    className="d-inline d-lg-none"
  >
    Pyr
    <span
      className="d-none d-md-inline"
    >
      omancy
    </span>
     
    1
  </span>
</span>
`;

exports[`<ResponsiveSchool /> with level: 2 should match the snapshot 1`] = `
<span>
  <span
    className="d-none d-lg-inline"
  >
    2nd
    -level 
    pyromancy
  </span>
  <span
    className="d-inline d-lg-none"
  >
    Pyr
    <span
      className="d-none d-md-inline"
    >
      omancy
    </span>
     
    2
  </span>
</span>
`;

exports[`<ResponsiveSchool /> with level: 3 should match the snapshot 1`] = `
<span>
  <span
    className="d-none d-lg-inline"
  >
    3rd
    -level 
    pyromancy
  </span>
  <span
    className="d-inline d-lg-none"
  >
    Pyr
    <span
      className="d-none d-md-inline"
    >
      omancy
    </span>
     
    3
  </span>
</span>
`;

exports[`<ResponsiveSchool /> with level: 4+ should match the snapshot 1`] = `
<span>
  <span
    className="d-none d-lg-inline"
  >
    4th
    -level 
    pyromancy
  </span>
  <span
    className="d-inline d-lg-none"
  >
    Pyr
    <span
      className="d-none d-md-inline"
    >
      omancy
    </span>
     
    4
  </span>
</span>
`;
